src/client/pmix_client_get.c

```diff
@@ -149,172 +134,18 @@ pmix_status_t PMIx_Get_nb(const pmix_proc_t *proc, const char *key,
-    }
-
-    /* not finding it is not an error - it could be in the
-     * modex hash table, so check it */
-    if (PMIX_SUCCESS == (rc = pmix_hash_fetch(&nptr->modex, proc->rank, key, &val))) {
-        pmix_output_verbose(2, pmix_globals.debug_output,
-                            "pmix: value retrieved from dstore");
-        /* need to push this into the event library to ensure
-         * the callback occurs within an event */
-        cb = PMIX_NEW(pmix_cb_t);
-        (void)strncpy(cb->nspace, nm, PMIX_MAX_NSLEN);
-        cb->rank = proc->rank;
-        cb->key = strdup(key);
-        cb->value_cbfunc = cbfunc;
-        cb->cbdata = cbdata;
-        /* pack the return data so the unpack routine can get it */
-        if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(&cb->data, val, 1, PMIX_VALUE))) {
-            PMIX_ERROR_LOG(rc);
-        }
-        /* cleanup */
-        if (NULL != val) {
-            PMIX_VALUE_RELEASE(val);
-        }
-        /* activate the event */
-        event_assign(&(cb->ev), pmix_globals.evbase, -1,
-                     EV_WRITE, getnb_shortcut, cb);
-        event_active(&(cb->ev), EV_WRITE, 1);
-        return PMIX_SUCCESS;
-    } else if (PMIX_ERR_NOT_FOUND == rc) {
-        /* we have the modex data from this proc, but didn't find the key
-         * the user requested. At this time, there is no way for the
-         * key to eventually be found, so all we can do is return
-         * the error */
-        pmix_output_verbose(2, pmix_globals.debug_output,
-                            "Error requesting key=%s for rank = %d, namespace = %s",
-                            key, proc->rank, nm);
-        return rc;
-    }
-
-  request:
-    /* if we got here, then we don't have the data for this proc. If we
-     * are a server, or we are a client and not connected, then there is
-     * nothing more we can do */
-    if (pmix_globals.server || (!pmix_globals.server && !pmix_globals.connected)) {
-        return PMIX_ERR_NOT_FOUND;
-    }
-
-    /* we also have to check the user's directives to see if they do not want
-     * us to attempt to retrieve it from the server */
-    for (n=0; n < ninfo; n++) {
-        if (0 == strcmp(info[n].key, PMIX_OPTIONAL) &&
-            info[n].value.data.flag) {
-            /* they don't want us to try and retrieve it */
-            pmix_output_verbose(2, pmix_globals.debug_output,
-                                "PMIx_Get key=%s for rank = %d, namespace = %s was not found - request was optional",
-                                key, proc->rank, nm);
-            return PMIX_ERR_NOT_FOUND;
-        }
-    }
-
-    /* Now we need to thread shift to the service
-     * thread for request processing */
+    /* thread-shift so we can check global objects */
     cb = PMIX_NEW(pmix_cb_t);
-    (void)strncpy(cb->nspace, nm, PMIX_MAX_NSLEN);
-    cb->rank = proc->rank;
-    cb->key = strdup(key);
+    cb->active = true;
+    cb->procs = (pmix_proc_t*)proc;
```

Comment:
@artpol84 That would definitely be contrary to standard practice for non-blocking functions. I've never seen any library that copied all input data on a non-blocking call - what we are doing here is the normal and expected.

I'm not sure we really want to protect against user errors - this could get really cumbersome if we do.